File: .github/copilot-instructions/examples/backend/php/UserServiceTest.php
<?php

declare(strict_types=1);

namespace App\Tests\Service;

use App\Service\UserService;
use App\Repository\UserRepositoryInterface;
use App\Security\PasswordHasherInterface;
use App\Exception\InvalidInputException;
use App\Exception\UserAlreadyExistsException;
use App\Exception\UserNotFoundException;
use App\DTO\UserDTO;
use App\DTO\CreateUserRequest;
use App\DTO\UpdateUserRequest;
use PHPUnit\Framework\TestCase;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\Attributes\DataProvider;

/**
 * Test suite for UserService - demonstrates PHPUnit best practices.
 *
 * Uses:
 * - PHPUnit 10+ with attributes (#[Test], #[DataProvider])
 * - Mocking with createMock()
 * - Named arguments
 * - Strict types
 * - Data providers for parameterized tests
 */
final class UserServiceTest extends TestCase
{
    private UserRepositoryInterface $userRepository;
    private PasswordHasherInterface $passwordHasher;
    private UserService $userService;

    protected function setUp(): void
    {
        $this->userRepository = $this->createMock(UserRepositoryInterface::class);
        $this->passwordHasher = $this->createMock(PasswordHasherInterface::class);
        
        $this->passwordHasher
            ->method('hash')
            ->willReturn('hashed_password_123');

        $this->userService = new UserService(
            userRepository: $this->userRepository,
            passwordHasher: $this->passwordHasher,
        );
    }

    #[Test]
    public function it_creates_user_with_valid_input(): void
    {
        // Arrange
        $request = new CreateUserRequest(
            username: 'johndoe',
            email: 'john@example.com',
            password: 'SecurePass123',
        );

        $expectedUser = new UserDTO(
            id: 1,
            username: 'johndoe',
            email: 'john@example.com',
            role: 'user',
        );

        $this->userRepository
            ->expects($this->once())
            ->method('findByEmail')
            ->with('john@example.com')
            ->willReturn(null);

        $this->userRepository
            ->expects($this->once())
            ->method('save')
            ->with(
                username: 'johndoe',
                email: 'john@example.com',
                passwordHash: 'hashed_password_123',
            )
            ->willReturn($expectedUser);

        // Act
        $result = $this->userService->createUser($request);

        // Assert
        $this->assertEquals($expectedUser, $result);
    }

    #[Test]
    public function it_throws_exception_when_user_already_exists(): void
    {
        // Arrange
        $request = new CreateUserRequest(
            username: 'johndoe',
            email: 'john@example.com',
            password: 'SecurePass123',
        );

        $existingUser = new UserDTO(
            id: 1,
            username: 'existing',
            email: 'john@example.com',
            role: 'user',
        );

        $this->userRepository
            ->method('findByEmail')
            ->willReturn($existingUser);

        // Assert & Act
        $this->expectException(UserAlreadyExistsException::class);
        $this->expectExceptionMessage('User with email john@example.com already exists');

        $this->userService->createUser($request);
    }

    #[Test]
    #[DataProvider('invalidUsernameProvider')]
    public function it_rejects_invalid_usernames(string $invalidUsername, string $expectedMessage): void
    {
        // Arrange
        $request = new CreateUserRequest(
            username: $invalidUsername,
            email: 'john@example.com',
            password: 'SecurePass123',
        );

        // Assert & Act
        $this->expectException(InvalidInputException::class);
        $this->expectExceptionMessage($expectedMessage);

        $this->userService->createUser($request);
    }

    public static function invalidUsernameProvider(): array
    {
        return [
            'too short' => ['ab', 'must be 3-30 characters'],
            'too long' => [str_repeat('a', 31), 'must be 3-30 characters'],
            'invalid chars' => ['john-doe', 'contain only letters'],
            'with space' => ['john doe', 'contain only letters'],
            'empty' => ['', 'required'],
        ];
    }

    #[Test]
    #[DataProvider('invalidEmailProvider')]
    public function it_rejects_invalid_emails(string $invalidEmail): void
    {
        // Arrange
        $request = new CreateUserRequest(
            username: 'johndoe',
            email: $invalidEmail,
            password: 'SecurePass123',
        );

        // Assert & Act
        $this->expectException(InvalidInputException::class);

        $this->userService->createUser($request);
    }

    public static function invalidEmailProvider(): array
    {
        return [
            'not an email' => ['notanemail'],
            'missing domain' => ['missing@domain'],
            'no local part' => ['@nodomain.com'],
            'with spaces' => ['spaces in@email.com'],
            'empty' => [''],
        ];
    }

    #[Test]
    #[DataProvider('weakPasswordProvider')]
    public function it_rejects_weak_passwords(string $weakPassword, string $expectedMessage): void
    {
        // Arrange
        $request = new CreateUserRequest(
            username: 'johndoe',
            email: 'john@example.com',
            password: $weakPassword,
        );

        // Assert & Act
        $this->expectException(InvalidInputException::class);
        $this->expectExceptionMessage($expectedMessage);

        $this->userService->createUser($request);
    }

    public static function weakPasswordProvider(): array
    {
        return [
            'too short' => ['short', 'at least 12 characters'],
            'no uppercase' => ['nouppercase123', 'uppercase letter'],
            'no lowercase' => ['NOLOWERCASE123', 'lowercase letter'],
            'no digits' => ['NoNumbersHere', 'one number'],
        ];
    }

    #[Test]
    public function it_authenticates_user_with_valid_credentials(): void
    {
        // Arrange
        $expectedUser = new UserDTO(
            id: 1,
            username: 'johndoe',
            email: 'john@example.com',
            role: 'user',
        );

        $this->userRepository
            ->method('findByEmail')
            ->with('john@example.com')
            ->willReturn($expectedUser);

        // Act
        $result = $this->userService->authenticate('john@example.com', 'password');

        // Assert
        $this->assertEquals($expectedUser, $result);
    }

    #[Test]
    public function it_returns_null_for_nonexistent_user(): void
    {
        // Arrange
        $this->userRepository
            ->method('findByEmail')
            ->willReturn(null);

        // Act
        $result = $this->userService->authenticate('nonexistent@example.com', 'password');

        // Assert
        $this->assertNull($result);
    }

    #[Test]
    public function it_gets_user_by_id(): void
    {
        // Arrange
        $expectedUser = new UserDTO(
            id: 123,
            username: 'johndoe',
            email: 'john@example.com',
            role: 'user',
        );

        $this->userRepository
            ->expects($this->once())
            ->method('findById')
            ->with(123)
            ->willReturn($expectedUser);

        // Act
        $result = $this->userService->getUser(123);

        // Assert
        $this->assertEquals($expectedUser, $result);
    }

    #[Test]
    public function it_returns_null_when_user_not_found(): void
    {
        // Arrange
        $this->userRepository
            ->method('findById')
            ->willReturn(null);

        // Act
        $result = $this->userService->getUser(999);

        // Assert
        $this->assertNull($result);
    }

    #[Test]
    public function it_updates_user_username(): void
    {
        // Arrange
        $existingUser = new UserDTO(
            id: 1,
            username: 'oldusername',
            email: 'john@example.com',
            role: 'user',
        );

        $request = new UpdateUserRequest(
            username: 'newusername',
            email: null,
        );

        $updatedUser = new UserDTO(
            id: 1,
            username: 'newusername',
            email: 'john@example.com',
            role: 'user',
        );

        $this->userRepository
            ->method('findById')
            ->willReturn($existingUser);

        $this->userRepository
            ->expects($this->once())
            ->method('update')
            ->with(
                userId: 1,
                username: 'newusername',
                email: 'john@example.com',
            )
            ->willReturn($updatedUser);

        // Act
        $result = $this->userService->updateUser(1, $request);

        // Assert
        $this->assertEquals('newusername', $result->username);
    }

    #[Test]
    public function it_throws_exception_when_updating_nonexistent_user(): void
    {
        // Arrange
        $request = new UpdateUserRequest(username: 'newusername', email: null);

        $this->userRepository
            ->method('findById')
            ->willReturn(null);

        // Assert & Act
        $this->expectException(UserNotFoundException::class);
        $this->expectExceptionMessage('User not found: 999');

        $this->userService->updateUser(999, $request);
    }

    #[Test]
    public function it_deletes_existing_user(): void
    {
        // Arrange
        $this->userRepository
            ->method('exists')
            ->with(1)
            ->willReturn(true);

        $this->userRepository
            ->expects($this->once())
            ->method('delete')
            ->with(1);

        // Act
        $this->userService->deleteUser(1);

        // Assert - expectations verified by PHPUnit
    }

    #[Test]
    public function it_throws_exception_when_deleting_nonexistent_user(): void
    {
        // Arrange
        $this->userRepository
            ->method('exists')
            ->willReturn(false);

        // Assert & Act
        $this->expectException(UserNotFoundException::class);
        $this->expectExceptionMessage('User not found: 999');

        $this->userService->deleteUser(999);
    }
}

File: .github/copilot-instructions/examples/backend/java/UserServiceTest.java
package com.example.demo.user;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.ArgumentCaptor;

import java.util.Optional;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;

/**
 * Test suite for UserService - demonstrates JUnit 5 best practices.
 *
 * Uses:
 * - JUnit 5
 * - Mockito for mocking
 * - AssertJ for fluent assertions
 * - Nested test classes for organization
 * - Parameterized tests
 * - DisplayName for readable test names
 */
@DisplayName("UserService")
class UserServiceTest {

    private UserRepository userRepository;
    private PasswordEncoder passwordEncoder;
    private UserService userService;

    @BeforeEach
    void setUp() {
        userRepository = mock(UserRepository.class);
        passwordEncoder = mock(PasswordEncoder.class);
        userService = new UserService(userRepository, passwordEncoder);

        // Default mock behavior
        when(passwordEncoder.encode(anyString())).thenReturn("hashed_password_123");
    }

    @Nested
    @DisplayName("createUser")
    class CreateUserTests {

        @Test
        @DisplayName("should create user with valid input")
        void shouldCreateUserWithValidInput() {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    "johndoe",
                    "john@example.com",
                    "SecurePass123"
            );

            var savedUser = User.builder()
                    .id(1L)
                    .username("johndoe")
                    .email("john@example.com")
                    .passwordHash("hashed_password_123")
                    .role(UserRole.USER)
                    .build();

            when(userRepository.existsByEmail(request.email())).thenReturn(false);
            when(userRepository.save(any(User.class))).thenReturn(savedUser);

            // Act
            var result = userService.createUser(request);

            // Assert
            assertThat(result)
                    .isNotNull()
                    .satisfies(user -> {
                        assertThat(user.id()).isEqualTo(1L);
                        assertThat(user.username()).isEqualTo("johndoe");
                        assertThat(user.email()).isEqualTo("john@example.com");
                        assertThat(user.role()).isEqualTo(UserRole.USER);
                    });

            verify(passwordEncoder).encode("SecurePass123");
            verify(userRepository).save(any(User.class));
        }

        @Test
        @DisplayName("should throw exception when user already exists")
        void shouldThrowExceptionWhenUserAlreadyExists() {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    "johndoe",
                    "john@example.com",
                    "SecurePass123"
            );

            when(userRepository.existsByEmail(request.email())).thenReturn(true);

            // Act & Assert
            assertThatThrownBy(() -> userService.createUser(request))
                    .isInstanceOf(UserService.UserAlreadyExistsException.class)
                    .hasMessageContaining("john@example.com");

            verify(userRepository, never()).save(any(User.class));
        }

        @ParameterizedTest
        @DisplayName("should reject invalid usernames")
        @ValueSource(strings = {"ab", "a".repeat(31), "john-doe", "john doe", ""})
        void shouldRejectInvalidUsernames(String invalidUsername) {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    invalidUsername,
                    "john@example.com",
                    "SecurePass123"
            );

            // Act & Assert
            assertThatThrownBy(() -> userService.createUser(request))
                    .isInstanceOf(UserService.InvalidInputException.class)
                    .hasMessageContainingAny("Username", "required");
        }

        @ParameterizedTest
        @DisplayName("should reject invalid emails")
        @ValueSource(strings = {"notanemail", "missing@domain", "@nodomain.com", "spaces in@email.com", ""})
        void shouldRejectInvalidEmails(String invalidEmail) {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    "johndoe",
                    invalidEmail,
                    "SecurePass123"
            );

            // Act & Assert
            assertThatThrownBy(() -> userService.createUser(request))
                    .isInstanceOf(UserService.InvalidInputException.class)
                    .hasMessageContainingAny("Email", "required");
        }

        @ParameterizedTest
        @DisplayName("should reject weak passwords")
        @CsvSource({
                "short, at least 12 characters",
                "nouppercase123, uppercase letter",
                "NOLOWERCASE123, lowercase letter",
                "NoNumbers, one number"
        })
        void shouldRejectWeakPasswords(String weakPassword, String expectedMessagePart) {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    "johndoe",
                    "john@example.com",
                    weakPassword
            );

            // Act & Assert
            assertThatThrownBy(() -> userService.createUser(request))
                    .isInstanceOf(UserService.InvalidInputException.class)
                    .hasMessageContaining(expectedMessagePart);
        }

        @Test
        @DisplayName("should save user with hashed password")
        void shouldSaveUserWithHashedPassword() {
            // Arrange
            var request = new UserService.CreateUserRequest(
                    "johndoe",
                    "john@example.com",
                    "SecurePass123"
            );

            var savedUser = User.builder()
                    .id(1L)
                    .username("johndoe")
                    .email("john@example.com")
                    .passwordHash("hashed_password_123")
                    .role(UserRole.USER)
                    .build();

            when(userRepository.existsByEmail(request.email())).thenReturn(false);
            when(userRepository.save(any(User.class))).thenReturn(savedUser);

            // Act
            userService.createUser(request);

            // Assert - Capture and verify the user entity
            ArgumentCaptor<User> userCaptor = ArgumentCaptor.forClass(User.class);
            verify(userRepository).save(userCaptor.capture());

            var capturedUser = userCaptor.getValue();
            assertThat(capturedUser.getPasswordHash()).isEqualTo("hashed_password_123");
            assertThat(capturedUser.getUsername()).isEqualTo("johndoe");
            assertThat(capturedUser.getEmail()).isEqualTo("john@example.com");
        }
    }

    @Nested
    @DisplayName("authenticate")
    class AuthenticateTests {

        @Test
        @DisplayName("should return user when credentials are valid")
        void shouldReturnUserWhenCredentialsAreValid() {
            // Arrange
            var user = User.builder()
                    .id(1L)
                    .username("johndoe")
                    .email("john@example.com")
                    .passwordHash("hashed_password")
                    .role(UserRole.USER)
                    .build();

            when(userRepository.findByEmail("john@example.com"))
                    .thenReturn(Optional.of(user));
            when(passwordEncoder.matches("password123", "hashed_password"))
                    .thenReturn(true);

            // Act
            var result = userService.authenticate("john@example.com", "password123");

            // Assert
            assertThat(result)
                    .isPresent()
                    .get()
                    .satisfies(userDTO -> {
                        assertThat(userDTO.id()).isEqualTo(1L);
                        assertThat(userDTO.username()).isEqualTo("johndoe");
                        assertThat(userDTO.email()).isEqualTo("john@example.com");
                    });
        }

        @Test
        @DisplayName("should return empty when user not found")
        void shouldReturnEmptyWhenUserNotFound() {
            // Arrange
            when(userRepository.findByEmail("nonexistent@example.com"))
                    .thenReturn(Optional.empty());

            // Act
            var result = userService.authenticate("nonexistent@example.com", "password123");

            // Assert
            assertThat(result).isEmpty();
        }

        @Test
        @DisplayName("should return empty when password is incorrect")
        void shouldReturnEmptyWhenPasswordIsIncorrect() {
            // Arrange
            var user = User.builder()
                    .id(1L)
                    .username("johndoe")
                    .email("john@example.com")
                    .passwordHash("hashed_password")
                    .role(UserRole.USER)
                    .build();

            when(userRepository.findByEmail("john@example.com"))
                    .thenReturn(Optional.of(user));
            when(passwordEncoder.matches("wrongpassword", "hashed_password"))
                    .thenReturn(false);

            // Act
            var result = userService.authenticate("john@example.com", "wrongpassword");

            // Assert
            assertThat(result).isEmpty();
        }
    }

    @Nested
    @DisplayName("getUser")
    class GetUserTests {

        @Test
        @DisplayName("should return user when found")
        void shouldReturnUserWhenFound() {
            // Arrange
            var user = User.builder()
                    .id(123L)
                    .username("johndoe")
                    .email("john@example.com")
                    .passwordHash("hashed_password")
                    .role(UserRole.USER)
                    .build();

            when(userRepository.findById(123L)).thenReturn(Optional.of(user));

            // Act
            var result = userService.getUser(123L);

            // Assert
            assertThat(result)
                    .isPresent()
                    .get()
                    .satisfies(userDTO -> {
                        assertThat(userDTO.id()).isEqualTo(123L);
                        assertThat(userDTO.username()).isEqualTo("johndoe");
                    });
        }

        @Test
        @DisplayName("should return empty when user not found")
        void shouldReturnEmptyWhenUserNotFound() {
            // Arrange
            when(userRepository.findById(999L)).thenReturn(Optional.empty());

            // Act
            var result = userService.getUser(999L);

            // Assert
            assertThat(result).isEmpty();
        }
    }

    @Nested
    @DisplayName("updateUser")
    class UpdateUserTests {

        @Test
        @DisplayName("should update username when provided")
        void shouldUpdateUsernameWhenProvided() {
            // Arrange
            var existingUser = User.builder()
                    .id(1L)
                    .username("oldusername")
                    .email("john@example.com")
                    .passwordHash("hashed_password")
                    .role(UserRole.USER)
                    .build();

            var request = new UserService.UpdateUserRequest("newusername", null);

            when(userRepository.findById(1L)).thenReturn(Optional.of(existingUser));
            when(userRepository.save(any(User.class))).thenAnswer(invocation -> invocation.getArgument(0));

            // Act
            var result = userService.updateUser(1L, request);

            // Assert
            assertThat(result.username()).isEqualTo("newusername");
            verify(userRepository).save(existingUser);
        }

        @Test
        @DisplayName("should throw exception when user not found")
        void shouldThrowExceptionWhenUserNotFound() {
            // Arrange
            var request = new UserService.UpdateUserRequest("newusername", null);
            when(userRepository.findById(999L)).thenReturn(Optional.empty());

            // Act & Assert
            assertThatThrownBy(() -> userService.updateUser(999L, request))
                    .isInstanceOf(UserService.UserNotFoundException.class)
                    .hasMessageContaining("999");
        }
    }

    @Nested
    @DisplayName("deleteUser")
    class DeleteUserTests {

        @Test
        @DisplayName("should delete existing user")
        void shouldDeleteExistingUser() {
            // Arrange
            when(userRepository.existsById(1L)).thenReturn(true);

            // Act
            userService.deleteUser(1L);

            // Assert
            verify(userRepository).deleteById(1L);
        }

        @Test
        @DisplayName("should throw exception when user not found")
        void shouldThrowExceptionWhenUserNotFound() {
            // Arrange
            when(userRepository.existsById(999L)).thenReturn(false);

            // Act & Assert
            assertThatThrownBy(() -> userService.deleteUser(999L))
                    .isInstanceOf(UserService.UserNotFoundException.class)
                    .hasMessageContaining("999");
        }
    }
}
